ee/clickhouse/models/cohort.py

```diff
@@ -163,6 +165,165 @@ def get_entity_cohort_subquery(cohort: Cohort, cohort_group: Dict, group_idx: in
+
+
+def performed_event_first_time_subquery(
+    prop: Property, team_id: int, prepend: Union[int, str]
+) -> Tuple[str, Dict[str, Any]]:
+    if prop.event_type == "action":
+        action_id = prop.event
+        event_id = None
+    else:
+        event_id = prop.event
+        action_id = None
+
+    days = prop.time_value
+
+    date_query, date_params = get_date_query(days, None, None)
+    entity_query, entity_params = _get_entity_query(event_id, action_id, team_id, prepend)
+    pdi_query = get_team_distinct_ids_query(team_id)
+    is_negation = prop.negation
+
+    extract_person = GET_PERSON_ID_FIRST_TIME_EVENT.format(
+        entity_query=entity_query, date_query=date_query, GET_TEAM_PERSON_DISTINCT_IDS=pdi_query,
+    )
+
+    params: Dict[str, Union[str, int]] = {**entity_params, **date_params, "team_id": team_id}
+
+    return f"{'NOT' if is_negation else ''} person_id IN ({extract_person})", params
+
+
+def get_periods(start_time, end_time, interval, interval_unit):
+    # TODO: do start and end time need truncating?
+    period_starts = []
+    params = {}
+    period_start = datetime.strptime(start_time, "%Y-%m-%d %H:%M:%S")
+    # TODO: much hacking
+    period_end = period_start + timedelta(**{f"{interval_unit}s": interval})
+    counter = 0
+    # TODO: handle complexities with incomplete periods
+    # Perhaps we shape the query such that you only give us start_time or end_time, and periods
+    # Did $pageview 10 times every week in the past month
+    # What's implicit above^ is that we're talking only in terms of periods, and an end_time.
+    # But its also ambigous, do 4 weeks, or 5 make up the past month? Depends on the month :$
+    while period_start < datetime.strptime(end_time, "%Y-%m-%d %H:%M:%S"):
+        period_starts.append(period_start)
+        # TODO: timezone nonsense, also depends on interval_unit. Check always utc + aware
+        params[f"period_{counter}"] = period_start.strftime("%Y-%m-%d")
+        period_start = period_end
+        period_end = period_start + timedelta(**{f"{interval_unit}s": interval})
+        counter += 1
+
+    #  Final boundary
+    params[f"period_{counter}"] = period_start.strftime("%Y-%m-%d")
+    # TODO: also pretty sure there's few edge cases I'm missing. Just getting shape out, test later
+    return len(period_starts), params
+
+
+def performed_event_lifecycle_subquery(
+    prop: Property, team_id: int, prepend: Union[int, str]
+) -> Tuple[str, Dict[str, Any]]:
+    if prop.event_type == "action":
+        action_id = prop.event
+        event_id = None
+    else:
+        event_id = prop.event
+        action_id = None
+
+    days = prop.time_value
+
+    date_query, date_params = get_date_query(days, None, None)
+    entity_query, entity_params = _get_entity_query(event_id, action_id, team_id, prepend)
+    pdi_query = get_team_distinct_ids_query(team_id)
+    is_negation = prop.negation
+
+    # TODO: which parameters determine period?
```

Comment:
wasn't sure what params mean what, so I basically winged it here and added TODOs.